### ISSUE TITLE
naive quantization error?

### ISSUE THREAD

**Issue by Slicerkao** [NervanaSystems/distiller#317]
Hi,
I think I get a somewhat basic question. 
In my experiment, I first pretrain my own vgg16 with batch norm on cifar10, which reaches acc 90%.
Then, I use the following code to quantize my model:

```
 # x_dat is the input data of a batch. 
 model.eval()
 quantizer = quantization.PostTrainLinearQuantizer(model, bits_activations=32, bits_parameters = 32)
 quantizer.prepare_model(distiller.get_dummy_input(dataset = None, device = None, input_shape = x_dat.shape)) 
```
out of my expectation, when I evaluate on the test data, the acc drop to 0.1. I expect that 32 bit quantization should leads to the original accuracy. When I try bits to be 8, the accuracy is almost the same as the original one. 

Am I ignoring any setup for this default quantizer ? 

****Update****
I've checked the weight distribution of the tensorboard, and I found that all the weight distribution are explosively scaled by a large number, while bias are not. Still not quite sure what's the reason, but I know that this (might ?) causes overflow in the subsequent operation like forward pass.

**Comment by levzlotnik** [comment]
Hi @Slicerkao,

When quantizing to 32 bit you're actually transforming the model to `int32` rather than `float32`, there the values are between `-2^31` and `2^31 -1`. When doing that, you're dividing the already rather small original range of activations (suppose it's `[-1,1]`) to chunks of size `2^32 -1` times **smaller**.   
This is why you see the weight distribution scaled so much, because the quantized weights are now integers (in floating point format) in the `[-2^31, 2^31 -1]` range (i.e. `~[-2e9, 2e9]`).
The bias is quantized+dequantized, so it stays in the original range.
Since we only simulate quantization but leave the quantized weights in floating point format (because pytorch doesn't fully support Quantized ops on GPU yet), this shouldn't affect the result any differently than 8 bit quantization.
I'm not quite sure myself why the accuracy drop happened. 
Could you please post the graph of the model after quantization?

Thanks,
Lev

**Comment by Slicerkao** [comment]
Hi,
Thanks for your kindly reply!

Actually, when I plot the intermediate output of Conv2d compared to its quantized version in the network, I figure that the quantized version is saturated to two floating point value, with one positive and one negative.

Also, I use cifar10 dataset with natural normalization, with vgg16bn loaded from torchvision. Otherwise, everything is pretty much default. I wonder if my use of your quantization function is correct? 

Thanks a lot for any further information. I’d be appreciated if my experiment can be reproduced.

**Comment by levzlotnik** [comment]
Hi @Slicerkao,

Could you please give more details about the saturation values? I'm not quite sure I understand...
Maybe it'll help if you post the script you're running...

Lev

**Comment by Slicerkao** [comment]
Hi @levzlotnik 

My script of running the quantization module is given as above. But I think I know what goes wrong:
when I use 32 bits for activation, 32 bits for weights, I should set the number of bits for accumulator ti some value higher than 32. Otherwise, it would clip off the intermediate value, causing the 'saturation' I observe at the quantized activation output

**Comment by guyjacob** [comment]
@Slicerkao - indeed the number of bits for the accumulator was almost certainly the reason for the massive degradation you saw. 

Your rule of thumb of 2 * (act_bits + wts_bits) is something you see a lot in HW. That is, for 8-bit activations and weights, you commonly see 32-bit accumulator being used.

It's not straight forward to calculate the size of the accumulator you actually need. You can do a worst-case calculation fairly easily, which depends on layer parameters. This is mentioned in our docs [here](https://nervanasystems.github.io/distiller/quantization.html) in the section "Avoiding Overflows". TL;DR - for a convolution layer with 'c' input channels and a kernel of size 'k_w * k_h', to **completely prevent overflows** you'd need an accumulator of size:

CEIL(2 * (act_bits + wts_bits) + log2(c * k_w * k_h))

This will make sure that even if all weights and activations are at their maximum possible value, you will not overflow. But you'll likely never encounter such a case in a real-world workload. So you could probably do with less, but how much less will really depend on the actual data. You could run a bunch of data through the model using a big accumulator and collect stats, and "trim" the accumulator according to those stats.

**Comment by guyjacob** [comment]
I had an error in the formula in the last comment, fixed now (shouldn't have multiplied act_bits+wts_bits by 2)

**Comment by guyjacob** [comment]
Closing this. @Slicerkao feel free to re-open, or open a new issue as needed.
